scalanet/src/io/iohk/scalanet/peergroup/dynamictls/ChannelAwareQueue.scala

```diff
@@ -0,0 +1,63 @@
+package io.iohk.scalanet.peergroup.dynamictls
+
+import io.iohk.scalanet.peergroup.CloseableQueue
+import io.netty.channel.ChannelConfig
+import monix.eval.Task
+import monix.execution.ChannelType
+
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+  * Wraps an underlying unbounded CloseableQueue queue and bounds it based on netty auto-read feature.
+  * While auto-read is disabled received messages start to accumulate in underlying os RCV tcp buffer. When RCV buffer is full
+  * sender SND buffer will start to buffer un-sent bytes. When sender SND buffer is full, the default behaviour is that
+  * write(xxx) will block.  In our case, sendMessage Task will not finish until there will be place in SND buffer
+  *
+  *
+  * WARNING: Actual limit may sometimes goes higher, as each netty read can return more than one message.
+  *
+  * @param limit how many items can accumulate in the queue
+  * @param queue is the underlying closeable message queue
+  */
+private[scalanet] final class ChannelAwareQueue[M] private (limit: Int, queue: CloseableQueue[M]) {
+  private val queueSize = new AtomicLong(0)
+
+  private val lowerBound: Int = Math.max(1, limit / 2)
+
+  def size: Long = queueSize.get()
+
+  def offer(channelConfig: ChannelConfig, a: M): Task[Unit] = {
+    Task(enableBPIfNecessary(channelConfig)).flatMap(_ => queue.offer(a).void)
+  }
+
+  def next(channelConfig: ChannelConfig): Task[Option[M]] = {
```

Comment:
Couldn't you move `channelConfig` into the constructor in the companion? Does it change during the liftetime? That would be bad, because it would mean that `offer` can disable one channel and `next` enable a different one, which doesn't make sense, does it?

Reply:
You are right, I just wanted avoid code reorganisation which you described in next comment. Potential for misuse is pretty small here as we are using this queue in one context only, so I thought we can live without it.  

But sure i can go that way 👍

Reply:
Yeah I think it's better to have the right abstraction and massage the code than admit that it's so wrecked that we just need to pile on.